blockchain/blockchain.go

```diff
@@ -301,14 +302,46 @@ func (b *Blockchain) SetL1Head(update *core.L1Head) error {
 		return err
 	}
 	return b.database.Update(func(txn db.Transaction) error {
-		return txn.Set(db.L1Height.Key(), updateBytes)
+		if err := txn.Set(db.L1Height.Key(), updateBytes); err != nil {
+			return err
+		}
+		header, err := blockHeaderByNumber(txn, update.BlockNumber)
+		switch {
+		case err == nil:
+			if header.Hash.Equal(update.BlockHash) {
+				return nil
+			}
+		case errors.Is(err, db.ErrKeyNotFound):
+			return nil
+		default:
+			return err
+		}
+
+		// New L1 head conflicts with L2 chain.
+		// Revert the L2 chain to one block before the new L1 head.
```

Comment:
I dont think this is enough.

We need to revert until the common ancestor of our L1 head and L2 chain.
For example;

Assume there are 2 forks; first one is our local L2 chain, which contains some reorged blocks( `B'X` )

`B0 <- B1 <-  B2 <-  B'3 <-  B'4 <-  B'5 <- B'6` and our L1 head is `B2`



From L1, we get verification on `B5` (since skipping `B3` & `B4` is perfectly valid), which is the canonical block for that slot.
We roll back to the chain below;

`B0 <- B1 <-  B2 <-  B'3 <-  B'4`

`B'3` and `B'4` are still in storage, but they are older then the new L1 head, `B6`. So we will never be able to revert them.

What we should do is, somehow, realize that `B2` is the most recent common block in both forks and revert until that.

Reply:
@joshklop I think the best way forward is to implement beacon-sync method. (ie start from L1 head and sync backwards).

That way, we don't have this problem at all.

Reply:
Agree that beacon sync is good in general, but does it help here? AFAIK, new L1 heads are not guaranteed to be in the L2 chain, so we still need to manage the case where they conflict with our current L2 chain.

Reply:
Yeap, it just makes it easier.

When the node first boots, it fetches the L1 head and starts syncing from that. So we for sure know that our first L1 head is compatible with the L2 chain that we are syncing to.

We can keep doing this for each incompatible L1 head that we get. We can just revert until the previous L1 head,  and beacon sync from new L1 head to the previous L1 head.

Reply:
Agree. I see your point. 

While making unrelated changes to accept only finalized L1 blocks, I was thinking about some simplifications we could make to the sync <> L1 integration; I think those would help here as well.

Will return to this once 0.4 and 0.12.1 changes are through.